### DISCUSSION TITLE
Are there plans to officially expose APIs for creating custom popups?

### DISCUSSION THREAD

**Discussion by zivarah** [NeogitOrg/neogit#1542] [Ideas]
Since git supports custom commands, it would be nice to be able to leverage the popup infrastructure in Neogit to provide popups for those commands as well. Whether it's your own personal scripts/aliases or some more organized project that adds git commands, providing a UI into those commands via Neogit would be awesome. I know this is possible in magit, and I think that in general it makes sense for Neogit to provide the platform for such extensions, similar to how Telescope is as much an infrastructure as it is a set of specific pickers.

Are there any plans to formalize and document the APIs needed to do so? Or is there some reason that Neogit maintainers would be hesitant to open things up for extension this way?

From some hacking about, I was able to put together a custom popup with very little effort in my personal nvim config. It seems like supporting this would really just come down to:

- Document the actual APIs that one is able to use
- Provide a way to put those custom popups under the `:Neogit` command
- ???

If the maintainers are open to it, I'd be happy to help make this a reality!

**Comment by CKolkey** [comment]
So, it totally is in my plans to expose this at some point - if you're keen on doing that I'd be happy to help however you need. The popup builder api has been stable for quite a while now, so I don't see an issue with making it public. Lemme know if you have any questions, otherwise, have at it :)

**Reply by zivarah** [reply]
I think my main question would be if there's any particular process you want to follow for documenting APIs. Would it make sense for now to just document which APIs are considered public in `docs/neogit.txt`, but refer folks to the source for documentation? That would avoid needing to double-document the APIs since they already have a fair amount of annotations in-source.

**Reply by CKolkey** [reply]
Yep, that makes sense to me. I think as long as any public methods have annotations, and the `opts` and `args` style table params have the fields enumerated for easy consumption via language server, thats a perfectly reasonable start.

**Comment by zivarah** [comment]
Closing this now that I've submitted https://github.com/NeogitOrg/neogit/pull/1546, any followup can happen there.
